### PR TITLE
[Function](array-type) support array_apply

### PR DESCRIPTION
**Pull request by eldenmoon**
Filter array to match specific binary condition

```
mysql> select array_apply([1000000, 1000001, 1000002], '=', 1000002);
+-------------------------------------------------------------+
| array_apply(ARRAY(1000000, 1000001, 1000002), '=', 1000002) |
+-------------------------------------------------------------+
| [1000002]                                                   |
+-------------------------------------------------------------+
```

# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

File: be/src/vec/functions/array/function_array_apply.cpp [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        } else {
+            src_column_data_ptr = check_and_get_column<ColumnNullable>(src_column)
+                                          ->get_nested_column()
+                                          .get_raw_data()
+                                          .data;
+        }
+        for (size_t i = 0; i < src_column.size(); ++i) {
+            T lhs_val = *reinterpret_cast<const T*>(src_column_data_ptr);
+            if (apply<T, op>(lhs_val, rhs_val)) {
+                column_filter_data[i] = 1;
```
Comment (xy720) [inline be/src/vec/functions/array/function_array_apply.cpp:130]: There is some enhancement space, dst offset can be calculated in here.

File: be/src/vec/functions/array/function_array_apply.cpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        } else if (which.is_decimal128()) {                                        \
+            *dst = _apply_internal<Decimal128, OP>(src_column, src_offsets, cmp);  \
+        } else if (which.is_decimal128i()) {                                       \
+            *dst = _apply_internal<Decimal128I, OP>(src_column, src_offsets, cmp); \
+        } else {                                                                   \
+            LOG(FATAL) << "unsupported type " << nested_type->get_name();          \
+        }                                                                          \
+    } while (0)
+
+    //
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_apply.cpp:191]:
```suggestion
```

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_apply.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+```sql
+array_apply(arr, op, val)
+```
+
+#### Arguments
+
+`arr` — The array to inspect.
+`op` — The compare operation
+`val` — The compared value
```
Comment (cambyzju) [inline docs/en/docs/sql-manual/sql-functions/array-functions/array_apply.md:47]: if value is null, we return what?

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_apply.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+#### Syntax
+
+```sql
+array_apply(arr, op, val)
+```
+
+#### Arguments
+
+`arr` — The array to inspect.
+`op` — The compare operation
```
Comment (cambyzju) [inline docs/en/docs/sql-manual/sql-functions/array-functions/array_apply.md:46]: which ops we supported?

File: regression-test/suites/query_p0/sql_functions/array_functions/test_array_with_scale_type.groovy [hunk 5]
```diff
@@ -68,5 +68,10 @@ suite("test_array_with_scale_type") {
         qt_select "select array(cast ('2022-12-02 22:23:24.999999' as datetimev2(3)),cast ('2022-12-02 22:23:23.997799' as datetimev2(3))) from ${tableName}"
         qt_select "select array(cast ('2022-12-02 22:23:24.999999' as datetimev2(3)),cast ('2022-12-02 22:23:23.997799' as datetimev2(3)))"
 
+        qt_select """select array_apply(c_array_datetimev2, "=", '2022-12-02 22:23:24.999999') from ${tableName}"""
+        qt_select """select array_apply(c_array_datetimev2, ">", '2022-12-01 22:23:24.999999') from ${tableName}"""
+        qt_select """select array_apply(c_array_decimal, "=", 22.678) from ${tableName}"""
+        qt_select """select array_apply(c_array_decimal, ">=", 22.1) from ${tableName}"""
```
Comment (cambyzju) [inline regression-test/suites/query_p0/sql_functions/array_functions/test_array_with_scale_type.groovy:74]: please add some cases about `array op null`

File: be/src/vec/functions/array/function_array_apply.cpp (hunk 1 above)
Comment (eldenmoon) [reply be/src/vec/functions/array/function_array_apply.cpp:130]: OK， but i the code will be little hard to understand to put offset caculated here, so i seperate them, I'll modify this if there is a performance issue

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_apply.md (hunk 3 above)
Comment (eldenmoon) [reply docs/en/docs/sql-manual/sql-functions/array-functions/array_apply.md:47]: null will be returned

File: regression-test/suites/query_p0/sql_functions/array_functions/test_array_with_scale_type.groovy (hunk 5 above)
Comment (eldenmoon) [reply regression-test/suites/query_p0/sql_functions/array_functions/test_array_with_scale_type.groovy:74]: ok

File: be/src/vec/functions/array/function_array_apply.cpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        }
+        for (size_t i = 0; i < src_column.size(); ++i) {
+            T lhs_val = *reinterpret_cast<const T*>(src_column_data_ptr);
+            if (apply<T, op>(lhs_val, rhs_val)) {
+                column_filter_data[i] = 1;
+            }
+            src_column_data_ptr += sizeof(T);
+        }
+        const IColumn::Filter& filter = column_filter_data;
+        ColumnPtr filtered = src_column.filter(filter, -1);
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_apply.cpp:135]: actually we know the result element number, we should better not use -1 here.

Comment (eldenmoon) [reply be/src/vec/functions/array/function_array_apply.cpp:135]: ok

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run p0

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**